lib/Optimizer/Transforms/LoopUnroll.cpp

```diff
@@ -194,6 +195,48 @@ class LoopUnrollPass : public cudaq::opt::impl::LoopUnrollBase<LoopUnrollPass> {
   }
 };
 
+/// After unrolling the loops, there may be duplicate registerName attributes in
+/// use. This pass will assign them unique names by appending a counter.
+class UpdateRegisterNamesPass
+    : public cudaq::opt::impl::UpdateRegisterNamesBase<
+          UpdateRegisterNamesPass> {
+public:
+  using UpdateRegisterNamesBase::UpdateRegisterNamesBase;
+
+  void runOnOperation() override {
+    auto *op = getOperation();
+
+    // First get a count of the number of times a variable is used.
+    std::unordered_map<std::string, int> varNameCounts;
+    op->walk([&](mlir::Operation *walkOp) {
+      if (auto prevAttr = walkOp->getAttr("registerName")) {
+        auto varName = prevAttr.cast<StringAttr>().getValue().str();
+        varNameCounts[varName]++;
```

Comment:
If this were a `DenseMap<StringRef, DenseSet<Operation*>>` , you could find all the operations with the same register names and compose sets of those operations. Then below, rather than re-traversing all of the Module, you could just iterate over the Sets in the Map and directly update a counter on each Operation there.

Reply:
I updated it to `DenseMap<StringRef, SmallVector<Operation *>>` in bf41f69. `SmallVector` was chosen to maintain ordering.